drivers/serial/uart_pl011.c

```diff
@@ -366,6 +376,81 @@ static int pl011_init(const struct device *dev)
 	return 0;
 }
 
+#ifdef CONFIG_EARLY_PRINT_USE_UART_PL011
+static inline int current_core_mmu_enabled(void)
+{
+#ifdef CONFIG_ARM64
+	return read_sctlr_el1() & SCTLR_M_BIT;
+#else
+	return -1;
+#endif
+}
+
+static inline volatile struct pl011_regs *const get_early_uart(const struct device *dev, int mmu)
+{
+	if (mmu) {
+		return (volatile struct pl011_regs *const)DEVICE_MMIO_GET(dev);
+	}
+#ifdef DEVICE_MMIO_IS_IN_RAM
+	return (volatile struct pl011_regs *const)(DEVICE_MMIO_ROM_PTR(dev)->phys_addr);
+#else
+	return (volatile struct pl011_regs *const)(DEVICE_MMIO_ROM_PTR(dev)->addr);
+#endif
+}
+
+int arch_printk_char_out(int c)
```

Comment:
Is this function supposed to be used by code? Who is the user?

Reply:
Oh, it's the `printk`'s default char out hook, https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/os/printk.c#L46-L55

Reply:
oh, TIL.